File: projects/sauce_final_test/pages/HomePage.ts
import { Page, Locator, expect } from '@playwright/test';
import { BasePage } from './BasePage';

export class HomePage extends BasePage {
    /**
     * This is the login page for the Swag Labs e-commerce website. Users can enter their username and password to access the product catalog.
     * URL Pattern: https://www.saucedemo.com/
     */

    /** Input field for the username. */
    get username_input(): Locator {
        return this.smartLocator({
            css: '[data-test=\'username\']',
            css: 'input#user-name',
        });
    }

    /** Input field for the password. */
    get password_input(): Locator {
        return this.smartLocator({
            css: '[data-test=\'password\']',
            css: 'input#password',
        });
    }

    /** Button to submit the login form. */
    get login_button(): Locator {
        return this.smartLocator({
            role: ['button', { name: '[data-test='login-button']' }],
            role: ['button', { name: 'input#login-button' }],
        });
    }

    /** Text displaying accepted usernames */
    get accepted_usernames_are_(): Locator {
        return this.smartLocator({
            text: 'text=Accepted usernames are:',
            text: 'div.login_credentials_wrap',
        });
    }

    /** Text displaying password for all users */
    get password_for_all_users_(): Locator {
        return this.smartLocator({
            text: 'text=Password for all users:',
            text: 'div.login_password',
        });
    }

    async verifyLoaded() {
        await expect(this.page).toHaveTitle(/.*HomePage.*/i);
        // Login button is present
        // Username input field is present
        // Password input field is present
    }
}

File: projects/boss_dsl_validation/pages/home_page.py
from playwright.async_api import Page, expect

class HomePage:
    """
    This is the login page for the Swag Labs e-commerce website. Users can enter their username and password to access the product catalog.
    URL Pattern: https://www.saucedemo.com/
    """
    def __init__(self, page: Page):
        self.page = page

    @property
    def username_input(self):
        """Input field for username."""
        return self.page.[data-test="username"].or_(self.page.input#user-name)

    @property
    def password_input(self):
        """Input field for password."""
        return self.page.[data-test="password"].or_(self.page.input#password)

    @property
    def login_button(self):
        """Button to submit login credentials."""
        return self.page.[data-test="login-button"].or_(self.page.input#login-button)

    @property
    def accepted_usernames_are_(self):
        """Text displaying accepted usernames."""
        return self.page.text=Accepted usernames are:.or_(self.page.div.login_credentials)

    @property
    def password_for_all_users_(self):
        """Text displaying password for all users."""
        return self.page.text=Password for all users:.or_(self.page.div.password)

    async def verify_loaded(self):
        """Executes critical checks to ensure page is ready."""
        await Page title is 'Swag Labs'
        await Username input field is present
        await Password input field is present
        await Login button is present
        await Accepted usernames text is present
        await Password for all users text is present